sql-plugin/src/main/scala/com/nvidia/spark/rapids/BatchWithPartitionData.scala

```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable.ArrayBuffer
+
+import ai.rapids.cudf.ColumnVector
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.RmmRapidsRetryIterator.withRetry
+import com.nvidia.spark.rapids.jni.SplitAndRetryOOM
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{StringType, StructType}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Wrapper class that specifies how many rows to replicate
+ * the partition value.
+ */
+case class PartitionRowData(rowValue: InternalRow, rowNum: Int)
+
+object PartitionRowData {
+  def from(rowValues: Array[InternalRow], rowNums: Array[Int]): Array[PartitionRowData] = {
+    rowValues.zip(rowNums).map {
+      case (rowValue, rowNum) => PartitionRowData(rowValue, rowNum)
+    }
+  }
+
+  def from(rowValues: Array[InternalRow], rowNums: Array[Long]): Array[PartitionRowData] = {
+    rowValues.zip(rowNums).map {
+      case (rowValue, rowNum) => PartitionRowData(rowValue, rowNum.toInt)
+    }
+  }
+}
+
+/**
+ * Class to wrap columnar batch and partition rows data and utility functions to merge them.
+ *
+ * @param inputBatch           Input ColumnarBatch.
+ * @param partitionedRowsData  Array of (row counts, InternalRow) pairs. Each entry specifies how
+ *                             many rows to replicate the partition value.
+ * @param partitionSchema      Schema of the partitioned data.
+ */
+case class BatchWithPartitionData(
+    inputBatch: SpillableColumnarBatch,
+    partitionedRowsData: Array[PartitionRowData],
+    partitionSchema: StructType) extends AutoCloseable {
+
+  /**
+   * Merges the partitioned data with the input ColumnarBatch.
+   *
+   * @return Merged ColumnarBatch.
+   */
+  def mergeWithPartitionData(): ColumnarBatch = {
+    withResource(inputBatch.getColumnarBatch()) { columnarBatch =>
+      if (partitionSchema.isEmpty) {
+        columnarBatch // No partition columns, return the original batch.
+      } else {
+        val partitionColumns = buildPartitionColumns()
```

Comment:
are these being closed if error happens? seems like should be a closeOnExcept around these unless I'm missing it.

Reply:
I missed closeOnExcept here. Added it.